pkg/monitor/monitor/monitor_manager.go

```diff
@@ -427,3 +422,196 @@ func (m *MonitorManager) patchTidbClusterStatus(tcRef *v1alpha1.TidbClusterRef,
+		return !monitor.Spec.Persistent, nil
+	}
+
+	if !errors.IsNotFound(err) {
+		klog.Errorf("Fail to get deployment for tm [%s/%s], err: %v", monitor.Namespace, monitor.Name, err)
+		return false, err
+	}
+	if !monitor.Spec.Persistent {
+		return true, nil
+	}
+	firstStsPvcName := GetMonitorFirstPVCName(monitor.Name)
+	deploymentPvcName := GetMonitorObjectName(monitor)
+	deploymentPvc, err := m.deps.PVCLister.PersistentVolumeClaims(monitor.Namespace).Get(deploymentPvcName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			klog.Errorf("Smooth migration for tm[%s/%s], get the PVC of the deployment error: %v", monitor.Namespace, monitor.Name, err)
+			return false, err
+		}
+
+		// If the PVC of the deployment does not exist and no old PV status, we don't need to migrate.
+		if monitor.Status.DeploymentStorageStatus == nil || len(monitor.Status.DeploymentStorageStatus.PvName) <= 0 {
+			return true, nil
+		}
+
+		deploymentPv, err := m.deps.PVLister.Get(monitor.Status.DeploymentStorageStatus.PvName)
+		if err != nil {
+			klog.Errorf("Smooth migration for tm[%s/%s], fail to patch PV %s, err: %v", monitor.Namespace, monitor.Name, monitor.Status.DeploymentStorageStatus.PvName, err)
+			return false, err
+		}
+
+		if deploymentPv.Spec.ClaimRef != nil && deploymentPv.Spec.ClaimRef.Name == firstStsPvcName {
+			// smooth migration successfully and clean status
+			monitor.Status.DeploymentStorageStatus = nil
+			return true, nil
+		}
+
+		err = m.patchPVClaimRef(deploymentPv, firstStsPvcName, monitor)
+		if err != nil {
+			klog.Errorf("Smooth migration for tm[%s/%s], fail to patch PV %s, err: %v", monitor.Namespace, monitor.Name, monitor.Status.DeploymentStorageStatus.PvName, err)
+			return false, err
+		}
+		// smooth migration successfully and clean status
+		monitor.Status.DeploymentStorageStatus = nil
+		return true, nil
+	}
+
+	if len(deploymentPvc.Spec.VolumeName) <= 0 {
+		klog.Infof("Smooth migration for tm[%s/%s], old pvc not bind pv and continue create statefulset", monitor.Namespace, monitor.Name)
+		return true, nil
+	}
+
+	deploymentPv, err := m.deps.PVLister.Get(deploymentPvc.Spec.VolumeName)
+	if err != nil {
+		klog.Errorf("Smooth migration for tm[%s/%s], fail to get PV %s, err: %v", monitor.Namespace, monitor.Name, deploymentPvc.Spec.VolumeName, err)
+		return false, err
+	}
+	if deploymentPv.Spec.PersistentVolumeReclaimPolicy == corev1.PersistentVolumeReclaimDelete {
+		klog.Errorf("Smooth migration for tm[%s/%s], pv[%s] policy is delete, it must be retain", monitor.Namespace, monitor.Name, deploymentPvc.Spec.VolumeName)
+		return false, err
+	}
+
+	if deploymentPv.Spec.ClaimRef != nil && deploymentPv.Spec.ClaimRef.Name == firstStsPvcName {
+		// smooth migration successfully and clean status
+		monitor.Status.DeploymentStorageStatus = nil
+		return true, nil
+	}
+
+	// firstly patch status
+	if monitor.Status.DeploymentStorageStatus == nil {
+		err = m.patchTidbMonitorStatus(monitor, deploymentPvc.Spec.VolumeName)
```

Comment:
If the status patch succeeds here but the following operations fail, e.g. in L528, then the status will be updated in `UpdateTidbMonitor` in `pkg/controller/tidbmonitor/tidb_monitor_control.go`, in this case, the patched status will be overwritten.